include/onnxruntime/core/session/onnxruntime_ep_c_api.h

```diff
@@ -465,6 +519,279 @@ struct OrtEpApi {
+                  _In_ int since_version_start, _In_ int since_version_end);
+
+  /** \brief Sets the name of the kernel's intended execution provider.
+   *
+   * \param[in] kernel_def_builder The OrtKernelDefBuilder instance.
+   * \param[in] ep_name A null-terminated string representing the execution provider's name.
+   *
+   * \snippet{doc} snippets.dox OrtStatus Return Value
+   *
+   * \since Version 1.24.
+   */
+  ORT_API2_STATUS(KernelDefBuilder_SetExecutionProvider, _In_ OrtKernelDefBuilder* kernel_def_builder,
+                  _In_ const char* ep_name);
+
+  /** \brief Sets the memory type for a kernel input.
+   *
+   * \param[in] kernel_def_builder The OrtKernelDefBuilder instance.
+   * \param[in] input_index The index of the input.
+   * \param[in] mem_type The input's memory type.
+   *
+   * \snippet{doc} snippets.dox OrtStatus Return Value
+   *
+   * \since Version 1.24.
+   */
+  ORT_API2_STATUS(KernelDefBuilder_SetInputMemType, _In_ OrtKernelDefBuilder* kernel_def_builder,
+                  _In_ size_t input_index, _In_ OrtMemType mem_type);
+
+  /** \brief Sets the memory type for a kernel output.
+   *
+   * \param[in] kernel_def_builder The OrtKernelDefBuilder instance.
+   * \param[in] output_index The index of the output.
+   * \param[in] mem_type The output's memory type.
+   *
+   * \snippet{doc} snippets.dox OrtStatus Return Value
+   *
+   * \since Version 1.24.
+   */
+  ORT_API2_STATUS(KernelDefBuilder_SetOutputMemType, _In_ OrtKernelDefBuilder* kernel_def_builder,
+                  _In_ size_t output_index, _In_ OrtMemType mem_type);
+
+  /** \brief Sets type constraints for a kernel argument represented as a string (e.g., "T").
+   *
+   * \param[in] kernel_def_builder The OrtKernelDefBuilder instance.
+   * \param[in] arg_name A null-terminated string representing the argument to constrain (e.g., "T").
+   * \param[in] types Array of OrtMLDataType instances representing allowed types for the argument.
+   *                  Must contain `num_types` elements.
+   * \param[in] num_types The number of OrtMLDataType elements in the `types` array.
+   *
+   * \snippet{doc} snippets.dox OrtStatus Return Value
+   *
+   * \since Version 1.24.
+   */
+  ORT_API2_STATUS(KernelDefBuilder_AddTypeConstraint, _In_ OrtKernelDefBuilder* kernel_def_builder,
+                  _In_ const char* arg_name, _In_reads_(num_types) const OrtMLDataType* const* types,
+                  _In_ size_t num_types);
+
+  /** \brief Creates a OrtKernelDef instance from the given kernel definition builder.
+   *
+   * \param[in] kernel_def_builder The OrtKernelDefBuilder instance.
+   * \param[out] kernel_def_out The new OrtKernelDef instance.
+   *
+   * \snippet{doc} snippets.dox OrtStatus Return Value
+   *
+   * \since Version 1.24.
+   */
+  ORT_API2_STATUS(KernelDefBuilder_Build, _In_ OrtKernelDefBuilder* kernel_def_builder,
+                  _Outptr_ OrtKernelDef** kernel_def_out);
+
```

Comment:
This PR does not yet add all KernelDefBuilder [functions](https://github.com/microsoft/onnxruntime/blob/be655f69f6c8623eebcac094626d0c8545951e6d/include/onnxruntime/core/framework/kernel_def_builder.h#L164). It's missing aliasing, "may inplace". However, the things that are missing may not be support common and could be added later.